Source/SwiftLintFramework/Rules/CompilerProtocolInitRule.swift

```diff
@@ -0,0 +1,156 @@
+//
+//  CompilerProtocolInitRule.swift
+//  SwiftLint
+//
+//  Created by Marcelo Fabri on 12/31/16.
+//  Copyright © 2016 Realm. All rights reserved.
+//
+
+import Foundation
+import SourceKittenFramework
+
+public struct CompilerProtocolInitRule: ASTRule, ConfigurationProviderRule {
+    public var configuration = SeverityConfiguration(.warning)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "compiler_protocol_init",
+        name: "Compiler Protocol Init",
+        description: "The initializers declared in compiler protocols such as `ExpressibleByArrayLiteral` " +
+                     "shouldn't be called directly.",
+        nonTriggeringExamples: [
+            "let set: Set<Int> = [1, 2]\n",
+            "let set = Set(array)\n"
+        ],
+        triggeringExamples: [
+            "let set = ↓Set(arrayLiteral: 1, 2)\n",
+            "let set = ↓Set.init(arrayLiteral: 1, 2)\n"
+        ]
+    )
+
+    public func validateFile(_ file: File, kind: SwiftExpressionKind,
+                             dictionary: [String : SourceKitRepresentable]) -> [StyleViolation] {
+        return violationRangesInFile(file, kind: kind, dictionary: dictionary).map {
+            StyleViolation(ruleDescription: type(of: self).description,
+                           severity: configuration.severity,
+                           location: Location(file: file, characterOffset: $0.location))
+        }
+    }
+
+    private func violationRangesInFile(_ file: File,
+                                       kind: SwiftExpressionKind,
+                                       dictionary: [String: SourceKitRepresentable]) -> [NSRange] {
+        guard kind == .call,
```

Comment:
There are a few things that could be improved in this function IMO. Especially since many variables are computed in every loop iteration even though they can't change.

Consider incorporating some of these changes:

```diff
         guard kind == .call,
-            let name = dictionary["key.name"] as? String else {
-                return []
+            let name = dictionary["key.name"] as? String,
+            let offset = (dictionary["key.offset"] as? Int64).flatMap({ Int($0) }),
+            let length = (dictionary["key.length"] as? Int64).flatMap({ Int($0) }) else {
+            return []
         }
 
+        let arguments = dictionary.enclosedArguments.flatMap({ $0["key.name"] as? String })
+
         for compilerProtocol in ExpressibleByCompiler.allProtocols {
-            guard compilerProtocol.initCallNames.contains(name),
-                case let arguments = dictionary.enclosedArguments.flatMap({ $0["key.name"] as? String }),
+            if compilerProtocol.initCallNames.contains(name),
                 compilerProtocol.matchArguments(arguments),
-                let offset = (dictionary["key.offset"] as? Int64).flatMap({ Int($0) }),
-                let length = (dictionary["key.length"] as? Int64).flatMap({ Int($0) }),
-                let range = file.contents.bridge().byteRangeToNSRange(start: offset, length: length) else {
-                    continue
+                let range = file.contents.bridge().byteRangeToNSRange(start: offset, length: length) {
+                return [range]
             }
-
-            return [range]
         }
```